src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,310 @@
+            public IEnumerator<KeyValuePair<TKey, Type>> GetEnumerator() => throw new NotSupportedException();
+            IEnumerator IEnumerable.GetEnumerator() => throw new NotSupportedException();
+        }
+
+        [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+        private sealed class DelayedType
+        {
+            private Type? _type;
+            public DelayedType(string assemblyName, string typeName)
+            {
+                AssemblyName = assemblyName;
+                TypeName = typeName;
+                _type = null;
+            }
+
+            public string AssemblyName { get; }
+            public string TypeName { get; }
+
+            public Type GetOrLoadType()
+            {
+                if (_type is null)
+                {
+                    lock (this)
+                    {
+                        Assembly targetAssembly = Assembly.Load(AssemblyName);
+                        _type = targetAssembly.GetType(TypeName, throwOnError: true)!;
+                    }
+                }
+                return _type;
+            }
+
+            public bool RepresentsSameType(DelayedType other)
+            {
+                return string.Equals(AssemblyName, other.AssemblyName, StringComparison.Ordinal)
+                        && string.Equals(TypeName, other.TypeName, StringComparison.Ordinal);
+            }
+        }
+
+        [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+        private sealed class LazyExternalTypeDictionary : LazyTypeLoadDictionary<string>
+        {
+            private static int ComputeHashCode(string key) => key.GetHashCode();
+
+            private readonly Dictionary<int, DelayedType> _lazyData = new();
+
+            protected override bool TryGetOrLoadType(string key, [NotNullWhen(true)] out Type? type)
+            {
+                int hash = ComputeHashCode(key);
+                if (!_lazyData.TryGetValue(hash, out DelayedType? value))
+                {
+                    type = null;
+                    return false;
+                }
+
+                type = value.GetOrLoadType();
+                return true;
+            }
+
+            public void Add(string key, TypeName parsedTarget)
+            {
+                int hash = ComputeHashCode(key);
+                if (_lazyData.ContainsKey(hash))
+                {
+                    ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException(key);
+                }
+
+                _lazyData.Add(hash, new DelayedType(parsedTarget.AssemblyName!.FullName, parsedTarget.FullName));
+            }
+        }
+
+        [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+        private sealed class LazyProxyTypeDictionary : LazyTypeLoadDictionary<Type>
+        {
+            // We don't include the assembly name for the hash code since it is not
```

Comment:
The FullName can include assembly names for generic components. Do we need to filter out the assembly names for the generic components here?

Reply:
Yes, we do. That is a going to be a complication, I think.

Reply:
Done.